### PR TITLE
CB-14399 : Added a new flow chain for recovering from DL Resize failures.

### PR DESCRIPTION
**Pull request by sxxgrc**
Jira: https://jira.cloudera.com/browse/CB-14399

This change is focused on introducing a new flow chain for recovering from DL resize failures. In creating this change I also fixed existing bugs which caused recovery to fail due to the way in which the detach logic functioned.

**Recovery Flow**

1. Detach the DL to be deleted.
2. Delete the DL.
3. Re-attach the old DL that has to be recovered.
4. Start the DL

There are therefore three main parts to this change:

- Creation of the DL Resize Recovery flow chain
     - Involves the new chain, new events, a trigger method, and a new flow within the existing SDX Detach flow (primarily in DetachActions)
- Modifying the SdxClusterStatusCheckerJob to work when a cluster's stack's CRN is modified (otherwise this would mark the cluster as failed during recovery)
- Modifying the Stack DB to have a constraint on unique (CRN, Terminated) pairs instead of just CRN (otherwise recovery would fail as both the resized DL and the old one have the same CRN)

I have tested this by running both my own unit tests, existing mock tests, and locally setting up a DL, resizing it, and then recovering it.

### CODE REVIEW

**Comment by sxxgrc** [comment]
Once this is unblocked, I plan on having Peter Erdos and Balazs Bihari review this as well to verify the SdxClusterStatusCheckerJob modifications.

**Comment by sxxgrc** [comment]
15446 has been merged so I will rebase this and open it up for reviews.

**Comment by sxxgrc** [comment]
@lacikaaa and @horadla23 I added you as reviewers since you looked at my PR for 15446. This is closely related so I would appreciate getting your input again.

**Comment by sxxgrc** [comment]
@sodre90 and @bbihari I added you both since I made a change to the logic in `SdxClusterStatusCheckerJob` and I know you both worked on that. I want to make sure my changes there look good.
Feel free to only look at that file and ignore the rest of the PR.

**Comment by sxxgrc** [comment]
Explanation for the schema change can be found in the Jira ticket comments.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:34]: When would this be true?

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:49]: How is it different from what is done in line 44?

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:34]: I think you mean to ask when would this be false since it is expected that recovery is done when the resized DL has failed creation or something else.

As to when it would be false, I'm not sure. I wanted to add more flexibility to the recovery process in case we wanted to retry recovery (after it already deleted the resized DL) or if we wanted to allow recovery before the resized DL is even created.

I am open to removing this though, it's not super important.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:49]: The `event.accepted` part is needed for the first event in a flow chain otherwise you get some errors and even a UI popup warning. 
The flow chain still runs but I figured we would not want warnings or errors to appear from the user's perspective.

File: datalake/src/main/java/com/sequenceiq/datalake/job/SdxClusterStatusCheckerJob.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/job/SdxClusterStatusCheckerJob.java:80]: Please add a comment with the explanation for the change.

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/job/SdxClusterStatusCheckerJob.java]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachFlowConfig.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachFlowConfig.java:82]: Is there a reason to have detach and reattach in the same flow?

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachFlowConfig.java:82]: Not a huge reason aside from simplicity since the recovery process is identical to the automatic recovery done in the detach flow. It didn't make sense to me to make an entire new flow if the flow step would be the same on both.

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachDetachUtils.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachDetachUtils.java:31]: What's wrong with the original implementation? Why is this refactor done?

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxAttachDetachUtils.java:31]: I modified this in order to add https://github.com/hortonworks/cloudbreak/pull/12228/files/d241d343c00b8a2d544d86ba3f51490d90ca7a4f#diff-4480719b7166ebd935c5cf33cb173cf2a4b6899df15f901679b078583cdb7af5R112

Specifically, I wanted to make sure that if recovery fails, everything is brought back to the point before recovery started (if possible).

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (foldik) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:35]: Can you please create a few methods in this class to instantiate these events. I think it would be cleaner to have methods like:
```java
SdxDeleteStartEvent deleteStartEvent(DatalakeResizeRecoveryFlowChainStartEvent event)

SdxStartDetachRecoveryEvent startDetachEvent(DatalakeResizeRecoveryFlowChainStartEvent event)

SdxStartStartEvent startDataLakeEvent(DatalakeResizeRecoveryFlowChainStartEvent event)
```

All those transformation logic can be hided this way.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (foldik) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:118]: I don't see this method been used anywhere. Is this feature not finished?

**Review by foldik — CHANGES_REQUESTED** [review]
I requested a few changes.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:118]: It's not that this is not finished, but that the goal of this ticket was to create the recovery flow chain as well as a trigger hook for it. This will be expanded upon in the future.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:49]: @foldik  I see that makes sense, thank you for providing the example. I'll clean this up.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java]: Done.

File: core/src/main/resources/schema/app/20220204202948_CB-14399_Modify_unique_CRN_constraint_to_unique_CRN_terminated_constraint.sql
Comment (lacikaaa) [inline core/src/main/resources/schema/app/20220204202948_CB-14399_Modify_unique_CRN_constraint_to_unique_CRN_terminated_constraint.sql:4]: this will potentially cause on issue, where it's possible there are multiple terminated stack with the same crn. Repository calls expect single stack as a result and would fail if multiple is returned

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:121]: nit: can we call this `userCrn` if it's not id but crn?

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:29]: could you add logs and unit test?

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:277]: remove the comment and refactor the code so it would be obvious without the comment
eg creating methods with the right naming

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:262]: is this feature available in prod? changing these stuff could cause a problem if a flow is running when we are rolling out a new version

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java:8]: could you add `toString` implementation?

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:11]: same, please ensure events has `toString` implementation which involves calling the parent class `toString` implementation too

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:41]: this should be the other way, `SdxDetachEvent.SDX_DETACH_RECOVERY_EVENT` should have this class passed in, and this shouldn't be overwritteni here

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:35]: this is a bit odd, and it I think the logic could be simplified/be more obvious if there was a different action/state for the failure with a specific selector/event for that case

File: datalake/src/main/java/com/sequenceiq/datalake/flow/start/event/SdxStartStartEvent.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/start/event/SdxStartStartEvent.java:16]: if the `selector()` is overwritten, it doesn't make any sense to expect a `selector` parameter

File: datalake/src/main/java/com/sequenceiq/datalake/job/SdxClusterStatusCheckerJob.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/job/SdxClusterStatusCheckerJob.java:78]: I'm not an expert in this job, but shouldn't this job be unscheduled, then scheduled with the new crn instead of this?
cc @topolyai5 @bbihari @foldik

**Review by lacikaaa — CHANGES_REQUESTED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (kkalvagadda1) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:118]: triggerSdxResizeRecovery should be triggered from ResizeRecoveryService. This PR should the invocation as well.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (kkalvagadda1) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java:262]: This feature is under entitlement and is nor available for customers yet.

File: core/src/main/resources/schema/app/20220204202948_CB-14399_Modify_unique_CRN_constraint_to_unique_CRN_terminated_constraint.sql
Comment (kkalvagadda1) [reply core/src/main/resources/schema/app/20220204202948_CB-14399_Modify_unique_CRN_constraint_to_unique_CRN_terminated_constraint.sql:4]: As resourceCrn's are uniques. The only situation where we see two stacks using the same CRN is the recovery from failed resize. When that happens, one of the stacks should have been terminated.

I agree, with the change we allowing the possibility of having multiple stacks with same CRN in terminated state.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java
Comment (kkalvagadda1) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java:15]: I'm sure the recovery that triggers this flow chain by sending the event would have information about the old datalake.

It should be implementation details. All the recovery API would about is the datalake that they want to recover. Based on the state if the datalake appropriate recovery mechanism would be triggered.

File: datalake/src/main/java/com/sequenceiq/datalake/job/SdxClusterStatusCheckerJob.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/job/SdxClusterStatusCheckerJob.java:78]: Yes I can do that, I think I was originally wanting this to always be up to date to make it a more correct approach, but I definitely see that there could be some unintended consequences with this, as well as limiting how we manage the SDX stuff.
I will schedule a new job which is incredibly simple and overrides any pre-existing job with the same DL ID.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:121]: So this use of `userId` is in basically every single other method within this file. Would you like me to change all of those as well?

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:29]: What tests? I already have a test which verifies that the correct events are added to the queue, is there something else you want here?

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:29]: Added logs though.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java:8]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:11]: There is no parent implementation for that.

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:11]: Added the tostring method though.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:41]: Cleaned up.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/start/event/SdxStartStartEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/start/event/SdxStartStartEvent.java:16]: Removed the override.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/DatalakeResizeRecoveryFlowChainStartEvent.java:15]: I don't understand what you mean.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:35]: While I understand your rationale here, I personally don't know how simple it would be to introduce an entirely different event directed to the same state as this event.
I also don't want to create a new action/state since it would contain the exact logic as the state here. They only differ in their final transitions.

I am open to making this a little cleaner, such as having a tracked boolean instead or some other logic, but I personally don't think adding more events and actions would keep this simple and understandable.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:118]: @kkalvagadda1 Dan's current change kind of mixes with this, should we not wait for either of these changes to be merged before making that modification? I don't want to run into merge conflicts.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachActions.java]: I removed the comment, I think the code is clear enough.

**Comment by sxxgrc** [comment]
Made some major changes/fixed. Main points:

- No longer modifying the SdxClusterStatusCheckerJob, instead I am making new jobs whenever I modify the stack CRN
- No longer modifying the stack DB, instead I am introducing a detach step for the resized DL right before it is deleted, this ensures there are no issues with reusing the name and CRN

I have tested this thoroughly on my end and things seem to be working well.

I will remove the requested reviews from those who no longer need to look at this.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeRecoveryFlowEventChainFactory.java:45]: Added a lot more testing recently, please take a look.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeFlowEventChainFactory.java
Comment (yg-gd) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/chain/DatalakeResizeFlowEventChainFactory.java:64]: nit: spaces

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (yg-gd) [inline datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:80]: What exactly is this jobservice used for?

**Review by yg-gd — APPROVED** [review]

File: datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java
Comment (kkalvagadda1) [reply datalake/src/main/java/com/sequenceiq/datalake/service/sdx/attach/SdxDetachService.java:80]: This schedules an SDX status-checked task to poll the status of the cluster resources and update the cluster state accordingly.

File: core/src/main/resources/schema/app/20220204202948_CB-14399_Modify_unique_CRN_constraint_to_unique_CRN_terminated_constraint.sql
Comment (sxxgrc) [reply core/src/main/resources/schema/app/20220204202948_CB-14399_Modify_unique_CRN_constraint_to_unique_CRN_terminated_constraint.sql]: No longer doing this, marking this as resolved.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/event/SdxStartDetachRecoveryEvent.java:30]: Okay I'll get this changed today, while hopefully reusing events and the like so as to not make this too complex. I'll retest everything again too.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachRecoveryActions.java
Comment (lacikaaa) [inline datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachRecoveryActions.java:91]: `String statusReason = Optional,ofNullable(exception.getMessage()).orElse("SDX detach recovery failed")`

Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/detach/SdxDetachRecoveryActions.java]: Done.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (topolyai5) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:118]: Hi, how could you guarantee that this pr will work if just call the trigger method? I think we must wait until you can test and call the trigger. It will easier to fix before the merge than after

**Review by topolyai5 — CHANGES_REQUESTED** [review]
This code is not called from code. The problems will be revealed after when someone calls the trigger.

File: datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java
Comment (sxxgrc) [reply datalake/src/main/java/com/sequenceiq/datalake/flow/SdxReactorFlowManager.java:118]: I tied in the recovery service.

**Review by topolyai5 — APPROVED** [review]

**Review by topolyai5 — CHANGES_REQUESTED** [review]
Could you please create an E2E  test and IT follow-up ticket and could you please handle with immediately prio?

**Comment by kkalvagadda1** [comment]
> Could you please create an E2E test and IT follow-up ticket and could you please handle with immediately prio? I will accept if the ticket is created and if you add the tests into the next release

Created [CB-16621](https://jira.cloudera.com/browse/CB-16621) for adding e2e tests for recovery.

**Review by topolyai5 — APPROVED** [review]

**Pull request merged**
